### PR TITLE
Add ros2 service info

### PR DESCRIPTION
**Pull request by leeminju531**
Related Issue https://github.com/ros2/ros2cli/issues/766

### [DEPENDENCY] to track works related to this: 
- rmw_cyclonedds (https://github.com/ros2/rmw_cyclonedds/pull/427)
- rmw_fastrtps (https://github.com/ros2/rmw_fastrtps/pull/641)
- rmw_connextdds (https://github.com/ros2/rmw_connextdds/pull/93)
- rmw (https://github.com/ros2/rmw/pull/334)
- rmw_implementation (https://github.com/ros2/rmw_implementation/pull/208)
- rcl (https://github.com/ros2/rcl/pull/1011)
- rclpy (https://github.com/ros2/rclpy/pull/1024)
- rclcpp(https://github.com/ros2/rclcpp/pull/2072)

Anticipated output
service_info

Signed-off-by: leeminju531 <dlalswn531@naver.com>

### CODE REVIEW

**Comment by mjcarroll** [comment]
Running a full CI from (https://gist.github.com/mjcarroll/9e6d17c5e4b5b5ac9f9364365145277b)

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17614/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12156/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18184/)

**Comment by leeminju531** [comment]
@mjcarroll friendly ping, could you renew CI ?

**Comment by fujitatomoya** [comment]
@leeminju531 is this still draft? i am happy to review those PRs.

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17807/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12372/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18463/)

File: ros2cli/test/test_daemon.py
Comment (audrow) [inline ros2cli/test/test_daemon.py:243]:
```suggestion
    assert 1 == daemon_node.count_services(TEST_SERVICE_NAME)

```
The file should have an empty line at the end.

File: ros2service/ros2service/verb/info.py
Comment (audrow) [inline ros2service/ros2service/verb/info.py:54]:
```suggestion
                  node.count_services(service_name), end='\n')

```

File: ros2service/ros2service/verb/info.py
Comment (audrow) [inline ros2service/ros2service/verb/info.py:1]:
```suggestion
# Copyright 2022 Open Source Robotics Foundation, Inc.
```

**Review by audrow — APPROVED** [review]
This looks good to me with green CI.

It looks like there may be some code-style errors still. I added suggestions for adding new lines at the end.

File: ros2service/ros2service/verb/info.py
Comment (leeminju531) [reply ros2service/ros2service/verb/info.py]: fixed

File: ros2service/ros2service/verb/info.py
Comment (leeminju531) [reply ros2service/ros2service/verb/info.py]: fixed

**Comment by leeminju531** [comment]
friendly ping @audrow , I think my tasks are completed, are there more steps?
 I'm waiting for merge :smiley::laughing:

**Comment by fujitatomoya** [comment]
@leeminju531 are you willing to implement this for https://github.com/ros2/rclcpp?

about https://github.com/ros2/rmw/pull/334#discussion_r1057986256, any thoughts? @audrow @gbiggs

File: ros2service/ros2service/verb/info.py
Comment (fujitatomoya) [inline ros2service/ros2service/verb/info.py:48]: if there are two types in `service_types`, printing format would be different?

```bash
###service_types =["hoge/fuga"]
Type: hoge/fuga

###service_types =["hoge/fuga", "foo/bar"]
Type: ['hoge/fuga', 'foo/bar']
```

probably it would be nice to have the same format for both?

Comment (leeminju531) [inline ros2service/ros2service/verb/info.py:48]: @fujitatomoya
 IMO, a service name has usually one type, and it's better not to be a list format at this time. 
From my perspective, what about adding the following phrase? 





```suggestion
            if len(service_types) > 1:
                        print('There are more than two types in the same service')
            print('Type: %s' % type_str, end='\n')
```
image

**Comment by schornakj** [comment]
I think it would be useful to connect this to #703, which is my version of this feature.

From my perspective it's important to have a version of this new verb which can be backported to Humble, since I'm working with commercial and industrial clients who can only use LTS releases of ROS 2. If a feature doesn't make it into an LTS release then these users will not be able to benefit from it until at least May 2024 when J-Turtle is released.

It's a good approach to extend the fundamental `rcl` API to add more functionality. However, I expect that it will not be possible to backport this API change plus the dependent RMW implementation changes to Humble, since these are ABI-breaking changes that require recompiling these packages.

Since the files you've added to `ros2service/verb` are very similar to the ones I added in my version, I think we could do something like this:
- Merge #703 and create a PR to backport it to Humble
- Merge the upstream dependencies of this PR
- Merge this PR to use your `count_clients` and `count_services` functions in the `info` verb.

@leeminju531 what do you think?

File: ros2service/ros2service/verb/info.py
Comment (schornakj) [reply ros2service/ros2service/verb/info.py:48]: It's possible for a service to have multiple types if there are multiple servers advertising with the same name, which is the main debugging situation where I've been using this feature. In that case it's definitely helpful to print all the types to help figure out which server is where.

**Comment by leeminju531** [comment]
@schornakj it sounds reasonable to me 😊

**Comment by leeminju531** [comment]
It looks like this PR is still pending.
If any additional work is needed before merging, please feel free to let me know

File: ros2service/ros2service/verb/info.py
Comment (fujitatomoya) [reply ros2service/ros2service/verb/info.py:48]: I think these can be added after this PR with `verbose` option aligned with `ros2 topic info -v`, i think that is @leeminju531 's assumption here. btw, which can be https://github.com/ros2/ros2cli/pull/703.

Comment (fujitatomoya) [reply ros2service/ros2service/verb/info.py:48]: i will take this comment back, since the current implementation aligned with `ros2 topic info`.

**Review by fujitatomoya — APPROVED** [review]

**Comment by fujitatomoya** [comment]
@leeminju531 all related PRs look good to me except minor comments. i think it would be probably better to rebase all these PRs, and the we can start the CI.

**Comment by leeminju531** [comment]
:ok_hand: All the things are addressed. @fujitatomoya

**Comment by fujitatomoya** [comment]
@leeminju531 thanks for rebasing and addressing comments.

CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19503/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14023/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20223/)

**Comment by fujitatomoya** [comment]
failing https://github.com/leeminju531/rmw_implementation/blob/59c7868449b283b38370d2e6c25aa9f3f94096ce/test_rmw_implementation/test/test_graph_api.cpp#L993.

passing next client count though, https://github.com/leeminju531/rmw_implementation/blob/59c7868449b283b38370d2e6c25aa9f3f94096ce/test_rmw_implementation/test/test_graph_api.cpp#L1007

probably we need to add some sleep to wait the graph update?

**Comment by leeminju531** [comment]
Yeah, I think adding a brief sleep can solve this.

I believe the failure can occur not only in Windows but in any operating system. The test attempts to discover the topic right after creating participants, so the time to discover was faster than its period.

**Comment by leeminju531** [comment]
The sleep added in [rmw_implementation](https://github.com/ros2/rmw_implementation/pull/208). could you run CI again? @fujitatomoya

**Comment by fujitatomoya** [comment]
@clalancette can you check https://github.com/ros2/rmw_implementation/pull/208? once that is approved, i will start the CI again.

**Comment by fujitatomoya** [comment]
@clalancette thanks for merging rmw packages. do you still need time for clients libraries including ros2cli? or i can start the CI based on the latest source.

**Comment by leeminju531** [comment]
Thanks for merging all dependencies, Could someone run CI for this ? 

Additionally, I've considered the '--verbose' option for details like `topic info -v`. However, it seems to require ABI breaking changes.  Perhaps it needs something similar to `rmw_get_clients_info_by_service` like [rmw_get_publishers_info_by_topic](https://github.com/ros2/rmw/blob/83445be486deae8c78d275e092eafb4bf380bd49/rmw/include/rmw/get_topic_endpoint_info.h#L26#L109). 

Nonetheless, it appears that information on which node owns the clients or services is available [here](https://github.com/ros2/ros2cli/pull/703). In my experience, displaying this information seems to be quite useful.

**Review by fujitatomoya — APPROVED** [review]

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19659/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14166/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20382/)

**Pull request merged**
